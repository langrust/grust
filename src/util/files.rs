use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}