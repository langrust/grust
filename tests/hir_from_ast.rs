use codespan_reporting::files::SimpleFiles;

use grustine::{error::display, hir_from_ast};

#[test]
fn hir_from_ast_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/counter.gr").expect("unkown file"),
    );

    match hir_from_ast(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/blinking.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_using_function.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/pid.gr").expect("unkown file"),
    );

    match hir_from_ast(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
