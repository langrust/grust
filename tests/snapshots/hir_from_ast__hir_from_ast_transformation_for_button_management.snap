---
source: tests/hir_from_ast.rs
expression: file
---
typedefs:
  - id: 19
    kind:
      Enumeration:
        elements:
          - 17
          - 18
    location: ~
  - id: 22
    kind:
      Enumeration:
        elements:
          - 20
          - 21
    location: ~
functions: []
nodes:
  - id: 26
    unscheduled_equations:
      25:
        id: 25
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing: ~
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 23
                      typing: ~
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: ~
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 5
                                typing: ~
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 2
                                            typing: ~
                                            location: ~
                                            dependencies: ~
                                          inputs:
                                            - kind:
                                                FollowedBy:
                                                  constant:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Constant:
                                                            constant:
                                                              Integer: 0
                                                    typing: ~
                                                    location: ~
                                                    dependencies: ~
                                                  expression:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 25
                                                    typing: ~
                                                    location: ~
                                                    dependencies: ~
                                              typing: ~
                                              location: ~
                                              dependencies: ~
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                                - kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 24
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                      typing: ~
                      location: ~
                      dependencies: ~
          typing: ~
          location: ~
          dependencies: ~
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph: ~
  - id: 35
    unscheduled_equations:
      31:
        id: 31
        expression:
          kind:
            Expression:
              expression:
                Match:
                  expression:
                    kind:
                      Expression:
                        expression:
                          Application:
                            function_expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 2
                              typing: ~
                              location: ~
                              dependencies: ~
                            inputs:
                              - kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 27
                                typing: ~
                                location: ~
                                dependencies: ~
                    typing: ~
                    location: ~
                    dependencies: ~
                  arms:
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 17
                        typing: ~
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Enumeration:
                                enum_id: 22
                                elem_id: 21
                        typing: ~
                        location: ~
                        dependencies: ~
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 18
                        typing: ~
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Application:
                                function_expression:
                                  kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 15
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                                inputs:
                                  - kind:
                                      Expression:
                                        expression:
                                          Application:
                                            function_expression:
                                              kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 11
                                              typing: ~
                                              location: ~
                                              dependencies: ~
                                            inputs:
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 33
                                                typing: ~
                                                location: ~
                                                dependencies: ~
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 29
                                                typing: ~
                                                location: ~
                                                dependencies: ~
                                    typing: ~
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Enumeration:
                                            enum_id: 22
                                            elem_id: 20
                                    typing: ~
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Enumeration:
                                            enum_id: 22
                                            elem_id: 21
                                    typing: ~
                                    location: ~
                                    dependencies: ~
                        typing: ~
                        location: ~
                        dependencies: ~
          typing: ~
          location: ~
          dependencies: ~
        location: ~
      32:
        id: 32
        expression:
          kind:
            Expression:
              expression:
                Match:
                  expression:
                    kind:
                      Expression:
                        expression:
                          Application:
                            function_expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 2
                              typing: ~
                              location: ~
                              dependencies: ~
                            inputs:
                              - kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 27
                                typing: ~
                                location: ~
                                dependencies: ~
                    typing: ~
                    location: ~
                    dependencies: ~
                  arms:
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 17
                        typing: ~
                        location: ~
                      - ~
                      - []
                      - kind:
                          FollowedBy:
                            constant:
                              kind:
                                Expression:
                                  expression:
                                    Constant:
                                      constant:
                                        Boolean: false
                              typing: ~
                              location: ~
                              dependencies: ~
                            expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 32
                              typing: ~
                              location: ~
                              dependencies: ~
                        typing: ~
                        location: ~
                        dependencies: ~
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 18
                        typing: ~
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Application:
                                function_expression:
                                  kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 11
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                                inputs:
                                  - kind:
                                      Expression:
                                        expression:
                                          Identifier:
                                            id: 33
                                    typing: ~
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Application:
                                            function_expression:
                                              kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 5
                                              typing: ~
                                              location: ~
                                              dependencies: ~
                                            inputs:
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 29
                                                typing: ~
                                                location: ~
                                                dependencies: ~
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 30
                                                typing: ~
                                                location: ~
                                                dependencies: ~
                                    typing: ~
                                    location: ~
                                    dependencies: ~
                        typing: ~
                        location: ~
                        dependencies: ~
          typing: ~
          location: ~
          dependencies: ~
        location: ~
      33:
        id: 33
        expression:
          kind:
            NodeApplication:
              node_id: 26
              inputs:
                - - 23
                  - kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 34
                    typing: ~
                    location: ~
                    dependencies: ~
                - - 24
                  - kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 28
                    typing: ~
                    location: ~
                    dependencies: ~
              output_id: 25
          typing: ~
          location: ~
          dependencies: ~
        location: ~
      34:
        id: 34
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 7
                    typing: ~
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 2
                                typing: ~
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 9
                                            typing: ~
                                            location: ~
                                            dependencies: ~
                                          inputs:
                                            - kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 27
                                              typing: ~
                                              location: ~
                                              dependencies: ~
                                            - kind:
                                                Expression:
                                                  expression:
                                                    Enumeration:
                                                      enum_id: 19
                                                      elem_id: 18
                                              typing: ~
                                              location: ~
                                              dependencies: ~
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                      typing: ~
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 2
                                typing: ~
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    FollowedBy:
                                      constant:
                                        kind:
                                          Expression:
                                            expression:
                                              Constant:
                                                constant:
                                                  Boolean: true
                                        typing: ~
                                        location: ~
                                        dependencies: ~
                                      expression:
                                        kind:
                                          Expression:
                                            expression:
                                              Application:
                                                function_expression:
                                                  kind:
                                                    Expression:
                                                      expression:
                                                        Identifier:
                                                          id: 9
                                                  typing: ~
                                                  location: ~
                                                  dependencies: ~
                                                inputs:
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 27
                                                    typing: ~
                                                    location: ~
                                                    dependencies: ~
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Enumeration:
                                                            enum_id: 19
                                                            elem_id: 17
                                                    typing: ~
                                                    location: ~
                                                    dependencies: ~
                                        typing: ~
                                        location: ~
                                        dependencies: ~
                                  typing: ~
                                  location: ~
                                  dependencies: ~
                      typing: ~
                      location: ~
                      dependencies: ~
          typing: ~
          location: ~
          dependencies: ~
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph: ~
component: ~
location: ~
